File: Components/WinForms/VB/ContactManagement/ContactDetail.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports DevExpress.Xpo

Namespace ContactManagement
	Public Partial Class ContactDetail
		Inherits System.Windows.Forms.Form
		Private Sub SetupBindings()
			txtOID.DataBindings.Add("Text", theContact, "Oid")
			txtFirstName.DataBindings.Add("Text", theContact, "FirstName")
			txtLastName.DataBindings.Add("Text", theContact, "LastName")
			txtPhoneNumber.DataBindings.Add("Text", theContact, "PhoneNumber")
			txtEmail.DataBindings.Add("Text", theContact, "Email")
			Dim collection As XPCollection = New XPCollection(theContact.Session, GetType(Company))
			collection.DisplayableProperties = "Name;This"
			cmbEmployer.DataSource = collection
			cmbEmployer.DisplayMember = "Name"
			cmbEmployer.ValueMember = "This"
			cmbEmployer.DataBindings.Add("SelectedValue", theContact, "Employer!")
		End Sub

		Public Property Contact() As Contact
			Get
				Return theContact
			End Get
			Set
				theContact = Value
				SetupBindings()
			End Set
		End Property

		Public Sub New()
			InitializeComponent()
		End Sub
		Public Sub New(ByVal theContact As Contact)
			Me.New()
			Me.Contact = theContact
		End Sub

		Private Sub btnAddresses_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnAddresses.Click
			Dim list As ObjectList = New ObjectList("Addresses", Me.Contact.Addresses, "Owner.DisplayName;Street;City;IsDefault")
			AddHandler list.OnInsertObject, AddressOf OnInsertAddress
			AddHandler list.OnEditObject, AddressOf OnEditAddress
			list.ShowDialog(Me)
		End Sub

		Private Sub btnDocs_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnDocs.Click
			Dim list As ObjectList = New ObjectList("Attachments", Me.Contact.Attachments, "Owner.DisplayName;Name")
			AddHandler list.OnInsertObject, AddressOf OnInsertAttachment
			AddHandler list.OnEditObject, AddressOf OnEditAttachment
			list.ShowDialog(Me)
		End Sub
		Private Sub OnEditAttachment(ByVal sender As Object, ByVal e As EditObjectEventArgs)
			Dim attachmentDetailForm As AttachmentDetail = New AttachmentDetail()
			attachmentDetailForm.Attachment = CType(e.ObjectToEdit, Attachment)
			If attachmentDetailForm.ShowDialog(Me) <> System.Windows.Forms.DialogResult.OK Then
				CType(e.ObjectToEdit, Attachment).Reload()
			End If
		End Sub
		Private Sub OnInsertAttachment(ByVal sender As Object)
			Dim attachmentDetailForm As AttachmentDetail = New AttachmentDetail()
			Dim attach As Attachment = New Attachment(Contact.Session)
			attach.Owner = Contact
			attachmentDetailForm.Attachment = attach
			If attachmentDetailForm.ShowDialog(Me) = System.Windows.Forms.DialogResult.OK Then
				Me.Contact.Attachments.Add(attach)
			End If
		End Sub
		Private Sub OnEditAddress(ByVal sender As Object, ByVal e As EditObjectEventArgs)
			Dim addressDetailForm As AddressDetail = New AddressDetail()
			addressDetailForm.Address = CType(e.ObjectToEdit, Address)
			If addressDetailForm.ShowDialog(Me) <> System.Windows.Forms.DialogResult.OK Then
				CType(e.ObjectToEdit, Address).Reload()
			End If
		End Sub
		Private Sub OnInsertAddress(ByVal sender As Object)
			Dim addressDetailForm As AddressDetail = New AddressDetail()
			Dim address As Address = New Address(Contact.Session)
			address.Owner = Contact
			addressDetailForm.Address = address
			If addressDetailForm.ShowDialog(Me) = System.Windows.Forms.DialogResult.OK Then
				Me.Contact.Addresses.Add(address)
			End If
		End Sub
	End Class
End Namespace

File: Components/WinForms/CS/ContactManagement/ContactDetail.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using DevExpress.Xpo;

namespace ContactManagement {
    public partial class ContactDetail : System.Windows.Forms.Form {
        private void SetupBindings() {
            txtOID.DataBindings.Add("Text", theContact, "Oid");
            txtFirstName.DataBindings.Add("Text", theContact, "FirstName");
            txtLastName.DataBindings.Add("Text", theContact, "LastName");
            txtPhoneNumber.DataBindings.Add("Text", theContact, "PhoneNumber");
            txtEmail.DataBindings.Add("Text", theContact, "Email");
            XPCollection collection = new XPCollection(theContact.Session, typeof(Company));
            collection.DisplayableProperties = "Name;This";
            cmbEmployer.DataSource = collection;
            cmbEmployer.DisplayMember = "Name";
            cmbEmployer.ValueMember = "This";
            cmbEmployer.DataBindings.Add("SelectedValue", theContact, "Employer!");
        }

        public Contact Contact {
            get { return theContact; }
            set {
                theContact = value;
                SetupBindings();
            }
        }

        public ContactDetail() {
            InitializeComponent();
        }
        public ContactDetail(Contact theContact)
            : this() {
            this.Contact = theContact;
        }
        
        private void btnAddresses_Click(object sender, System.EventArgs e) {
            ObjectList list = new ObjectList("Addresses", this.Contact.Addresses,
                "Owner.DisplayName;Street;City;IsDefault");
            list.OnInsertObject += new InsertObjectEventHandler(this.OnInsertAddress);
            list.OnEditObject += new EditObjectEventHandler(this.OnEditAddress);
            list.ShowDialog(this);
        }

        private void btnDocs_Click(object sender, System.EventArgs e) {
            ObjectList list = new ObjectList("Attachments", this.Contact.Attachments,
                "Owner.DisplayName;Name");
            list.OnInsertObject += new InsertObjectEventHandler(this.OnInsertAttachment);
            list.OnEditObject += new EditObjectEventHandler(this.OnEditAttachment);
            list.ShowDialog(this);
        }
        private void OnEditAttachment(object sender, EditObjectEventArgs e) {
            AttachmentDetail attachmentDetailForm = new AttachmentDetail();
            attachmentDetailForm.Attachment = (Attachment)e.ObjectToEdit;
            if(attachmentDetailForm.ShowDialog(this) != DialogResult.OK)
                ((Attachment)e.ObjectToEdit).Reload();
        }
        private void OnInsertAttachment(object sender) {
            AttachmentDetail attachmentDetailForm = new AttachmentDetail();
            Attachment attach = new Attachment(Contact.Session);
            attach.Owner = Contact;
            attachmentDetailForm.Attachment = attach;
            if(attachmentDetailForm.ShowDialog(this) == DialogResult.OK) {
                this.Contact.Attachments.Add(attach);
            }
        }
        private void OnEditAddress(object sender, EditObjectEventArgs e) {
            AddressDetail addressDetailForm = new AddressDetail();
            addressDetailForm.Address = (Address)e.ObjectToEdit;
            if(addressDetailForm.ShowDialog(this) != DialogResult.OK)
                ((Address)e.ObjectToEdit).Reload();
        }
        private void OnInsertAddress(object sender) {
            AddressDetail addressDetailForm = new AddressDetail();
            Address address = new Address(Contact.Session);
            address.Owner = Contact;
            addressDetailForm.Address = address;
            if(addressDetailForm.ShowDialog(this) == DialogResult.OK) {
                this.Contact.Addresses.Add(address);
            }
        }
    }
}
